File: 999_pos/trunk/scripts/console.js
/**
 * Library with the Console class for displaying messages.
 * @package Client
 * @author Roberto Oliveros
 */

/**
 * Constructor function.
 * @param string sConsole
 */
function Console(sConsole){
	this._mElementDiv = document.getElementById(sConsole);
}

/**
 * Displays the message in the div element.
 */
Console.prototype.displayMessage = function(sMsg){
	this._mElementDiv.innerHTML += sMsg;
	this._mElementDiv.scrollTop = this._mElementDiv.scrollHeight;
}

/**
 * For displaying any kind of failure on the console div and right next to the element.
 * @param string sMsg
 * @param string sElementId
 */
Console.prototype.displayFailure = function(sMsg, sElementId){
	var newP = '<p id="failed-' + sElementId + '" class="failure">' + sMsg + '</p>';
	var oElement = document.getElementById(sElementId + '-failed');
	
	// Display the message and indicate which property cause the failure.
	this.displayMessage(newP);
	oElement.className = 'failed';
}

/**
* For displaying the errors.
* @param string sMsg
*/
Console.prototype.displayError = function(sMsg){
	var elementP = document.getElementById('error');
	// If there was a message.
	if(elementP)
		this._mElementDiv.removeChild(elementP);
	
	var newP = '<p id="error" class="error">' + sMsg + '</p>';
	// Display the error message.
	this.displayMessage(newP);
}

/**
 * Removes the failure message from the console and from the element that provoked.
 * @param string sElementId
 */
Console.prototype.cleanFailure = function(sElementId){
	var elementP = document.getElementById('failed-' + sElementId);
	
	// If there was a message.
	if(elementP){
		this._mElementDiv.removeChild(elementP);
		this._mElementDiv.scrollTop = this._mElementDiv.scrollHeight;
	
		var oElement = document.getElementById(sElementId + '-failed');
		oElement.className = 'hidden';
	}
}
 
/**
 * Clean the element div from all messages. Elements fields as well.
 */
Console.prototype.reset = function(){
	var arrElements = this._mElementDiv.getElementsByTagName('*');
	while(arrElements.length > 0){
		var sId = arrElements[0].getAttribute('id'); 
		if(sId != 'error')
			this.cleanFailure(sId.substring(sId.indexOf('-') + 1));
		else
			this._mElementDiv.removeChild(arrElements[0]);
	}
}

File: 999_exe/trunk/console/console.cpp
/*
 * console.cpp
 *
 *  Created on: 27/04/2010
 *      Author: pc
 */

#include "console.h"

/**
 * @class Console
 * Use to display messages to the user.
 */

/**
 * Sets the console QWebFrame for displaying messages.
 */
void Console::setFrame(QWebFrame *frame)
{
	m_Div = frame->findFirstElement("#console");
}

/**
 * Displays a validation failure in the console.
 */
void Console::displayFailure(QString msg, QString elementId)
{
	QString newP = "<p id=\"failed-" + elementId + "\" class=\"failure\">"
			+ msg + "</p>";

	showElementIndicator(elementId);

	displayMessage(newP);
}

/**
 * Removes a validation failure from the console.
 */
void Console::cleanFailure(QString elementId)
{
	QWebElement elementP = m_Div.findFirst("#failed-" + elementId);

	// If there was a message.
	if (!elementP.isNull()) {
		elementP.removeFromDocument();
		m_Div.evaluateJavaScript("this.scrollTop = this.scrollHeight;");
		hideElementIndicator(elementId);
	}
}

/**
 * Displays an error message on the div.
 */
void Console::displayError(QString msg)
{
	QWebElement elementP = m_Div.findFirst("#error");
	// If there was a message.
	if (!elementP.isNull())
		elementP.removeFromDocument();

	QString newP = "<p id=\"error\" class=\"error\">" + msg + "</p>";
	displayMessage(newP);
}

/**
 * Cleans the console and hides all the element indicators.
 */
void Console::reset()
{
	QWebElementCollection collection = m_Div.findAll("p");

	foreach (QWebElement element, collection) {
		QString id = element.attribute("id");
		if (id != "error" && id != "info") {
			cleanFailure(id.mid(id.indexOf("-") + 1));
		} else {
			element.removeFromDocument();
		}
	}
}

/**
 * Displays the message on the div.
 */
void Console::displayMessage(QString msg)
{
	m_Div.appendInside(msg);
	m_Div.evaluateJavaScript("this.scrollTop = this.scrollHeight;");
}
